### PR TITLE
infra(eslint): limit usage of globals in src

### LINKED ISSUE #3145: Prevent usage of features not broadly available
Related to #3143

- #3143

We should adjust our build or lint tooling to report bad usage of globals/features that aren't compatible with both modern browsers and node JS versions.

See here for an initial attempt:

- https://github.com/faker-js/faker/pull/3144#discussion_r1775884619

### PR DESCRIPTION
**Pull request by ST-DDT**
Fixes #3145

- #3145

Ban the usage of globals in our `src` folder.
All usages of these globals have to be marked with a eslint disable rule making them easier to spot.

### CODE REVIEW

File: eslint.config.ts [hunk 1]
```diff
@@ -200,6 +200,17 @@ const config: ReturnType<typeof tseslint.config> = tseslint.config(
...
 
   //#region overrides
+  {
+    // Must override the config from typescript-eslint
+    files: ['src/**/*.ts'],
+    rules: {
+      'no-undef': 'error',
+    },
+    languageOptions: {
+      // Don't allow any globals in our TypeScript files
```
Comment (ST-DDT) [inline eslint.config.ts]:
```suggestion
      // Don't allow any globals in our TypeScript files - unless explicitly ignored
```

File: eslint.config.ts [hunk 2]
```diff
@@ -200,6 +200,17 @@ const config: ReturnType<typeof tseslint.config> = tseslint.config(
...
+      'no-undef': 'error',
+    },
+    languageOptions: {
+      // Don't allow any globals in our TypeScript files - unless explicitly ignored
+      globals: {},
+    },
+  },
   {
     files: ['src/**/*.ts'],
     rules: {
```
Comment (Shinigami92) [inline eslint.config.ts]: **question:** why not merging these both together? they addressing the same file glob

Comment (ST-DDT) [reply eslint.config.ts]: Fixed. (The config originally looked different and I only simplified it without checking for existing groups)

**Review by Shinigami92 — APPROVED** [review]

**Comment by xDivisionByZerox** [comment]
What happened to the original implementation where globals that are not specific to an environment were allowed?

**Comment by ST-DDT** [comment]
> What happened to the original implementation where globals that are not specific to an environment were allowed?

It was never added to our source and this solution was recommended to me by members of eslint.

**Comment by ST-DDT** [comment]
Also AFAIK we don't actually want to use any additional globals anyway.

**Review by xDivisionByZerox — APPROVED** [review]

**Pull request merged**